src/Grant/GrantFactory.php

```diff
@@ -0,0 +1,89 @@
+<?php
+
+namespace League\OAuth2\Client\Grant;
+
+class GrantFactory
+{
+    /**
+     * @var array
+     */
+    protected $registry = [];
+
+    /**
+     * Define a grant class in the registry.
+     *
+     * @param  string $name
+     * @param  string $class
+     * @return $this
+     */
+    public function setGrant($name, $class)
+    {
+        $this->registry[$name] = $this->checkGrant($class);
+
+        return $this;
+    }
+
+    /**
+     * Get a grant instance by name.
+     *
+     * If the grant has not be registered, a default grant will be loaded.
+     *
+     * @param  string $name
+     * @param  string $options
+     * @return GrantInterface
+     */
+    public function getGrant($name, array $options = [])
+    {
+        if (empty($this->registry[$name])) {
+            $this->registry[$name] = $this->getGrantClass($name);
+        }
+
+        $class = $this->registry[$name];
+
+        return new $class($options);
```

Comment:
Due to the way grants operate right now, there's no harm in having singletons. I just have a small worry that someone will create a custom grant type and make it stateful.